### PR TITLE
Prefer ansible_facts namespace in examples

### PR DESCRIPTION
**Pull request by netzvieh**
##### SUMMARY

We should prefer the use of `ansible_facts` in the examples. Also remove one mention of dot vs. array notation in `faq.rst` since it has it's own explanation a few lines down.

##### ISSUE TYPE

- Docs Pull Request

### CODE REVIEW

**Comment by caphrim007** [comment]
I have no issue with this for the BIG-IP modules. @wojtek0806 to backport

**Review by bcoca — APPROVED** [review]

File: docs/docsite/rst/user_guide/playbooks_filters.rst [hunk 1]
```diff
@@ -1239,7 +1239,7 @@ To format a date using a string (like with the shell date command), use the "str
     {{ '%H:%M:%S' | strftime }}
 
     # Use ansible_date_time.epoch fact
-    {{ '%Y-%m-%d %H:%M:%S' | strftime(ansible_date_time.epoch) }}
+    {{ '%Y-%m-%d %H:%M:%S' | strftime(ansible_facts['date_time'['epoch']) }}
```
Comment (dagwieers) [inline docs/docsite/rst/user_guide/playbooks_filters.rst]: Although I like `{{ '%Y-%m-%d %H:%M:%S' | strftime(ansible_facts.date_time.epoch) }}` more.
```suggestion
    {{ '%Y-%m-%d %H:%M:%S' | strftime(ansible_facts['date_time']['epoch']) }}
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
I am more in favor of the dot-notation for many examples as it reads better than additional brackets.

File: docs/docsite/rst/user_guide/playbooks_filters.rst (hunk 1 above)
Comment (netzvieh) [reply docs/docsite/rst/user_guide/playbooks_filters.rst]: Added the missing ]. Thanks!

For the brackets vs. dots notation I used what I saw in other places like [0]. And dot notation has some caveats, though they shouldn't be relevant to the facts generated by ansible.

[0] https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html?highlight=ansible_facts

**Comment by felixfontein** [comment]
> We should prefer the use of ansible_facts in the examples.

Out of curiosity: *why* should be prefer that?

**Comment by dagwieers** [comment]
My personal view on this is that we should move to a *built-in* `facts` variable so that Ansible facts become available through `facts.hostname`. And only advertise this once we are at the end of this *evolutionary-path-of-improvement*. But there's no consensus where exactly we are going, so it makes it impossible to come to a concise, practical and well-designed language/syntax IMO.

**Comment by netzvieh** [comment]
@felixfontein I'm basically following the [Ansible 2.5 Porting Guide](https://docs.ansible.com/ansible/latest/porting_guides/porting_guide_2.5.html#ansible-fact-namespacing) here.

> Ansible facts, which have historically been written to names like ansible_* in the main facts namespace, have been placed in their own new namespace, ansible_facts.* For example, the fact ansible_distribution is now best queried through the variable structure ansible_facts.distribution.
> 
> A new configuration variable, inject_facts_as_vars, has been added to ansible.cfg. Its default setting, ‘True’, keeps the 2.4 behavior of facts variables being set in the old ansible_* locations (while also writing them to the new namespace). This variable is expected to be set to ‘False’ in a future release. When inject_facts_as_vars is set to False, you must refer to ansible_facts through the new ansible_facts.* namespace.

**Comment by felixfontein** [comment]
Ok, that makes sense. But why use the array notation? It's pretty ugly, and also the changelog you refer to doesn't use it.

**Comment by bcoca** [comment]
The array notation also avoids several issues:
for example `mydict.keys` (returns all keys in a dictionary) is not the same as `mydict['keys']` (returns value assigne to 'keys' key)  as the dot notation is implemented by Jinja via a fallback mechanism to access the dictionaries, but if you use a key with the same name as an existing method, you will invoke the method and NOT the key , returning unexpected results.

**Comment by netzvieh** [comment]
ready_for_review

**Comment by netzvieh** [comment]
I rebased the branch.

@dagwieers would you be okay with the bracket/array notation?

I also think some `facts.hostname`/`facts['hostname']` notation would be neat. Thing is, as of now the documented examples are not consistent. And I think they should try to be, even if in the future the notation will switch to something different. And having a consistent syntax in the docs helps when the change to some different notation comes eventually as we can easily search for `ansible_facts` and replace the occurences. This would be more work now, as we would have to look for all `ansible_` variables and weed out those, that should still be called that way.

**Comment by felixfontein** [comment]
There's one problem with array notation which dot notation doesn't have: it doesn't play nicely with `default()` and the `defined` test. If `test == {}` (or is not defined), `test.foo.bar | default('')` works, while `test['foo']['bar'] | default('')` results in an exception.

(This could be fixed by also adding `__getitem__` to `AnsibleUndefined`, similar to `__getattr__`: https://github.com/ansible/ansible/blob/devel/lib/ansible/template/__init__.py#L201-L211)

**Comment by jillr** [comment]
Thanks very much for this PR @netzvieh, it's sparked a lot of [discussion](https://meetbot.fedoraproject.org/ansible-meeting/2019-09-10/ansible_core_irc_public_meeting.2019-09-10-19.00.log.html) about how folks use vars in different scenarios.  Unfortunately we've determined that while we would like to be able to be more consistent in the docs, we're unable to be prescriptive without running into lots of cases where either dot notation or bracket notation would be a problem.  So the best thing we can do is continue to support mixed examples in the docs, letting docs authors decide what the proper syntax is for what they're writing.

**Comment by s-hertel** [comment]
As @jillr mentioned, we agreed against removing mixed examples, but we would gladly accept a PR just updating the documentation to prefer `ansible_facts`.

**Pull request closed without merging**